File: java/com/google/copybara/buildozer/BuildozerOptions.java
/*
 * Copyright (C) 2016 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.copybara.buildozer;

import static java.nio.charset.StandardCharsets.UTF_8;

import com.beust.jcommander.Parameter;
import com.beust.jcommander.Parameters;
import com.google.common.base.Joiner;
import com.google.common.base.Preconditions;
import com.google.common.base.Splitter;
import com.google.common.collect.ImmutableList;
import com.google.common.collect.Lists;
import com.google.common.flogger.FluentLogger;
import com.google.copybara.GeneralOptions;
import com.google.copybara.Option;
import com.google.copybara.WorkflowOptions;
import com.google.copybara.exception.ValidationException;
import com.google.copybara.format.BuildifierOptions;
import com.google.copybara.util.BadExitStatusWithOutputException;
import com.google.copybara.util.CommandOutput;
import com.google.copybara.util.CommandOutputWithStatus;
import com.google.copybara.util.console.Console;
import com.google.copybara.util.console.Consoles;
import com.google.copybara.shell.Command;
import com.google.copybara.shell.CommandException;
import java.nio.file.Path;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/** Specifies how Buildozer is executed. */
@Parameters(separators = "=")
public final class BuildozerOptions implements Option {

  private final GeneralOptions generalOptions;
  private final BuildifierOptions buildifierOptions;
  private final WorkflowOptions workflowOptions;

  public BuildozerOptions(GeneralOptions generalOptions,
      BuildifierOptions buildifierOptions, WorkflowOptions workflowOptions) {
    this.generalOptions = Preconditions.checkNotNull(generalOptions);
    this.buildifierOptions = Preconditions.checkNotNull(buildifierOptions);
    this.workflowOptions = workflowOptions;
  }

  private static final FluentLogger logger = FluentLogger.forEnclosingClass();
  private static final Pattern targetNotFound =
      Pattern.compile(".*error while executing commands \\[.+\\] on target (?<error>.* not found)");

  @Parameter(names = "--buildozer-bin",
      description = "Binary to use for buildozer (Default is /usr/bin/buildozer)",
      hidden = true)
  public String buildozerBin = "/usr/bin/buildozer";

  private void logError(Console console, CommandOutput output) {
    Consoles.errorLogLines(console, "buildozer stdout: ", output.getStdout());
    Consoles.errorLogLines(console, "buildozer stderr: ", output.getStderr());
  }

  static class BuildozerCommand {

    private final List<String> targets;
    private final String cmd;

    BuildozerCommand(List<String> targets, String cmd) {
      this.targets = Preconditions.checkNotNull(targets);
      this.cmd = Preconditions.checkNotNull(cmd);
    }

    BuildozerCommand(String targets, String cmd) {
      this.targets = ImmutableList.of(Preconditions.checkNotNull(targets));
      this.cmd = Preconditions.checkNotNull(cmd);
    }

    @Override
    public String toString() {
      return cmd + "|" + Joiner.on('|').join(targets);
    }
  }

  /**
   * Runs buildozer with the given commands.
   */
  void run(Console console, Path checkoutDir, Iterable<BuildozerCommand> commands)
      throws ValidationException, TargetNotFoundException {
    List<String> args = Lists.newArrayList(
        buildozerBin, "-buildifier=" + buildifierOptions.buildifierBin);

    // We only use -k in keep going mode because it shows less errors (http://b/69386431)
    if (workflowOptions.ignoreNoop) {
      args.add("-k");
    }
    args.add("-f");
    args.add("-");
    try {
      Command cmd =
          new Command(
              args.toArray(new String[0]), /*environmentVariables*/ null, checkoutDir.toFile());
      CommandOutputWithStatus output = generalOptions.newCommandRunner(cmd)
          .withVerbose(generalOptions.isVerbose())
          .withInput(Joiner.on('\n').join(commands).getBytes(UTF_8))
          .execute();
      if (!output.getStdout().isEmpty()) {
        logger.atInfo().log("buildozer stdout: %s", output.getStdout());
      }
      if (!output.getStderr().isEmpty()) {
        logger.atInfo().log("buildozer stderr: %s", output.getStderr());
      }
    } catch (BadExitStatusWithOutputException e) {
      // Don't print the output for common/known errors.
      if (generalOptions.isVerbose()) {
        logError(console, e.getOutput());
      }
      if (e.getResult().getTerminationStatus().getExitCode() == 3) {
        // Buildozer exits with code == 3 when the build file was not modified and no output
        // was generated. This happens with expressions that match multiple targets, like
        // :%java_library
        throw new TargetNotFoundException(
            commandsMessage("Buildozer could not find a target for", commands));
      }
      if (e.getResult().getTerminationStatus().getExitCode() == 2) {
        ImmutableList<String> errors =
            Splitter.on('\n').splitToList(e.getOutput().getStderr()).stream()
                .filter(s -> !(s.isEmpty() || s.startsWith("fixed ")))
                .collect(ImmutableList.toImmutableList());
        ImmutableList.Builder<String> notFoundMsg = ImmutableList.builder();
        boolean allNotFound = true;
        for (String error : errors) {
          Matcher matcher = targetNotFound.matcher(error);
          if (matcher.matches()) {
            notFoundMsg.add(
                String.format("Buildozer could not find a target for %s", matcher.group(1)));
          } else if (error.contains("no such file or directory")
              || error.contains("not a directory")) {
            notFoundMsg.add("Buildozer could not find build file: " + error);
          } else {
            allNotFound = false;
          }
        }
        if (allNotFound) {
          throw new TargetNotFoundException(Joiner.on("\n").join(notFoundMsg.build()));
        }
      }
      // Otherwise we have already printed above.
      if (!generalOptions.isVerbose()) {
        logError(console, e.getOutput());
      }
      throw new ValidationException(String.format(
          "%s\nCommand stderr:%s",
          commandsMessage("Failed to execute buildozer with args", commands),
          e.getOutput().getStderr()),
          e);
    } catch (CommandException e) {
      String message = String.format("Error '%s' running buildozer command: %s",
          e.getMessage(), e.getCommand().toDebugString());
      console.error(message);
      throw new ValidationException(message, e);
    }
  }

  private static String commandsMessage(final String prefix, Iterable<BuildozerCommand> commands) {
    return prefix + ":\n  " + Joiner.on("\n  ").join(commands);
  }
}

File: src/Copybara.Core/Buildozer/BuildozerOptions.cs
/*
 * Copyright (C) 2016 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

using System.Collections.Immutable;
using System.Text;
using System.Text.RegularExpressions;
using Copybara.Common;
using Copybara.Exceptions;
using Copybara.Format;
using Copybara.Util;
using Copybara.Util.Console;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Logging.Abstractions;
using Command = Copybara.Util.Command;
using CommandException = Copybara.Util.CommandException;
using Console = Copybara.Util.Console.Console;

namespace Copybara.Buildozer;

/// <summary>Specifies how Buildozer is executed.</summary>
public sealed class BuildozerOptions : IOption
{
    private static readonly ILogger Logger = NullLogger.Instance;

    private static readonly Regex TargetNotFound = new(
        @".*error while executing commands \[.+\] on target (?<error>.* not found)",
        RegexOptions.Compiled);

    private readonly GeneralOptions _generalOptions;
    private readonly BuildifierOptions _buildifierOptions;
    private readonly WorkflowOptions _workflowOptions;

    public BuildozerOptions(
        GeneralOptions generalOptions,
        BuildifierOptions buildifierOptions,
        WorkflowOptions workflowOptions)
    {
        _generalOptions = Preconditions.CheckNotNull(generalOptions);
        _buildifierOptions = Preconditions.CheckNotNull(buildifierOptions);
        _workflowOptions = workflowOptions;
    }

    [Flag(
        "--buildozer-bin",
        "Binary to use for buildozer (Default is /usr/bin/buildozer)",
        Hidden = true)]
    public string BuildozerBin { get; set; } = "/usr/bin/buildozer";

    private void LogError(Console console, CommandOutput output)
    {
        Consoles.ErrorLogLines(console, "buildozer stdout: ", output.GetStdout());
        Consoles.ErrorLogLines(console, "buildozer stderr: ", output.GetStderr());
    }

    public sealed class BuildozerCommand
    {
        private readonly IReadOnlyList<string> _targets;
        private readonly string _cmd;

        internal BuildozerCommand(IReadOnlyList<string> targets, string cmd)
        {
            _targets = Preconditions.CheckNotNull(targets);
            _cmd = Preconditions.CheckNotNull(cmd);
        }

        internal BuildozerCommand(string targets, string cmd)
        {
            _targets = ImmutableArray.Create(Preconditions.CheckNotNull(targets));
            _cmd = Preconditions.CheckNotNull(cmd);
        }

        public override string ToString() => _cmd + "|" + string.Join('|', _targets);
    }

    /// <exception cref="ValidationException"/>
    /// <exception cref="TargetNotFoundException"/>
    internal void Run(Console console, string checkoutDir, IEnumerable<BuildozerCommand> commands)
    {
        string unused = RunCaptureOutput(console, checkoutDir, commands);
    }

    /// <summary>Runs buildozer with the given commands.</summary>
    /// <exception cref="ValidationException"/>
    /// <exception cref="TargetNotFoundException"/>
    internal string RunCaptureOutput(
        Console console, string checkoutDir, IEnumerable<BuildozerCommand> commands)
    {
        var commandList = commands.ToList();
        var args = new List<string>
        {
            BuildozerBin,
            "-buildifier=" + _buildifierOptions.BuildifierBin,
        };

        // We only use -k in keep going mode because it shows less errors (http://b/69386431)
        if (_workflowOptions.IgnoreNoop)
        {
            args.Add("-k");
        }
        args.Add("-f");
        args.Add("-");
        try
        {
            var cmd = new Copybara.Util.Command(args.ToArray(), null, checkoutDir);
            CommandOutputWithStatus output = _generalOptions.NewCommandRunner(cmd)
                .WithVerbose(_generalOptions.IsVerbose())
                .WithInput(Encoding.UTF8.GetBytes(string.Join('\n', commandList)))
                .Execute();
            if (output.GetStdout().Length != 0)
            {
                Logger.LogInformation("buildozer stdout: {Stdout}", output.GetStdout());
            }
            if (output.GetStderr().Length != 0)
            {
                Logger.LogInformation("buildozer stderr: {Stderr}", output.GetStderr());
            }
            return output.GetStdout();
        }
        catch (BadExitStatusWithOutputException e)
        {
            // Don't print the output for common/known errors.
            if (_generalOptions.IsVerbose())
            {
                LogError(console, e.GetOutput());
            }
            if (e.GetResult().TerminationStatus.GetExitCode() == 3)
            {
                // Buildozer exits with code == 3 when the build file was not modified and no output
                // was generated. This happens with expressions that match multiple targets, like
                // :%java_library
                throw new TargetNotFoundException(
                    CommandsMessage("Buildozer could not find a target for", commandList));
            }
            if (e.GetResult().TerminationStatus.GetExitCode() == 2)
            {
                var errors = e.GetOutput().GetStderr()
                    .Split('\n')
                    .Where(s => !(s.Length == 0 || s.StartsWith("fixed ", StringComparison.Ordinal)))
                    .ToList();
                var notFoundMsg = new List<string>();
                bool allNotFound = true;
                foreach (string error in errors)
                {
                    Match matcher = TargetNotFound.Match(error);
                    if (matcher.Success)
                    {
                        notFoundMsg.Add(
                            $"Buildozer could not find a target for {matcher.Groups["error"].Value}");
                    }
                    else if (error.Contains("no such file or directory")
                             || error.Contains("not a directory"))
                    {
                        notFoundMsg.Add("Buildozer could not find build file: " + error);
                    }
                    else
                    {
                        allNotFound = false;
                    }
                }
                if (allNotFound)
                {
                    throw new TargetNotFoundException(string.Join('\n', notFoundMsg));
                }
            }
            // Otherwise we have already printed above.
            if (!_generalOptions.IsVerbose())
            {
                LogError(console, e.GetOutput());
            }
            throw new ValidationException(
                string.Format(
                    "{0}\nCommand stderr:{1}",
                    CommandsMessage("Failed to execute buildozer with args", commandList),
                    e.GetOutput().GetStderr()),
                e);
        }
        catch (CommandException e)
        {
            string message = string.Format(
                "Error '{0}' running buildozer command: {1}",
                e.Message,
                e.GetCommand());
            console.Error(message);
            throw new ValidationException(message, e);
        }
    }

    private static string CommandsMessage(string prefix, IEnumerable<BuildozerCommand> commands) =>
        prefix + ":\n  " + string.Join("\n  ", commands);
}
